lib/LanguageServerRename/Model/Renamer.php

```diff
@@ -0,0 +1,314 @@
+            count($locations)
+        ));
+
+        return $this->locationsToWorkspaceEdit($locations, $oldName, $newName);
+    }
+
+    private function documentAndPositionToNodeAndOffset(TextDocumentItem $textDocument, Position $position): array
+    {
+        $offset = PositionConverter::positionToByteOffset($position, $textDocument->text);
+        
+        $rootNode = $this->parser->parseSourceFile($textDocument->text);
+        $node = $rootNode->getDescendantNodeAtPosition($offset->toInt());
+        return [$offset, $node];
+    }
+
+    private function locationsToWorkspaceEdit(array $locations, string $oldName, string $newName): WorkspaceEdit
+    {
+        // group locations by document
+        $locationsByUri = [];
+        foreach ($locations as $location) {
+            /** @var Location $location */
+            $uri = (string)$location->uri();
+            if (!isset($locationsByUri[$uri])) {
+                $locationsByUri[$uri] = [];
+            }
+            $locationsByUri[$uri][] = $location;
+        }
+
+        $documentEdits = [];
+        foreach ($locationsByUri as $uri => $locations) {
+            list($textEdits, $renameOperation) = $this->documentLocationsToTextEdits($uri, $locations, $oldName, $newName);
+            $documentEdits[] = new TextDocumentEdit(
+                new VersionedTextDocumentIdentifier($uri, $this->getDocumentVersion($uri)),
+                $textEdits
+            );
+            if($renameOperation !== null)
+                $documentEdits[] = $renameOperation;
+        }
+
+        return new WorkspaceEdit(null, $documentEdits);
+    }
+
+    private function documentLocationsToTextEdits(string $documentUri, array $locations, string $oldName, string $newName): array
+    {
+        $edits = [];
+        $rename = null;
+        $documentContent = $this->getDocumentText($documentUri);
+        $rootNode = $this->parser->parseSourceFile($documentContent);
+        foreach ($locations as $location) {
+            /** @var Location $location */
+            $node = $rootNode->getDescendantNodeAtPosition($location->offset()->toInt());
+            
+            if(($r = $this->renameFileIfNeeded($node, $documentUri, $oldName, $newName)) !== null){
+                $rename = $r;
+            }
+
+            $nodeNameText = $this->nodeUtils->getNodeNameText($node, $documentContent);
+            if ($nodeNameText !== $oldName) {
+                continue;
+            }
+            
+            $position = $this->nodeUtils->getNodeNameStartPosition($node, $oldName);
+            
+            if ($position !== null) {
+                $edits[] = new TextEdit(
+                    new Range(
+                        $position,
+                        new Position($position->line, $position->character + mb_strlen($oldName))
+                    ),
+                    $newName
+                );
+            }
+        }
+
+        return [ $edits, $rename ];
```

Comment:
if we remove renmae file then this becomes returns only TextEdits

Reply:
That is going to be reworked completely (we'd be returning `RenameResult`s) so no need to clear that now, I think.